File: ios/Runner/Database/Extensions/DatabaseDecoderExtension.swift
//
//  DatabaseDecoderExtension.swift
//  Runner
//
//  Created by Hashim Khan on 24/01/2024.
//


import SQLite3

extension DatabaseManager {
    
    
    
    // Helper to extract the JSON string from a given column index
    func extractJsonString(from statement: OpaquePointer?, at index: Int32) -> String? {
        guard let columnText = sqlite3_column_text(statement, index) else {
            return nil
        }
        return String(cString: columnText)
    }
    
    // Helper to extract the string from a given column index
    func extractString(from statement: OpaquePointer?, at index: Int32) -> String? {
        guard let columnText = sqlite3_column_text(statement, index) else {
            return nil
        }
        return String(cString: columnText)
    }
    
    
    // Helper to extract the int from a given column index
    func extractInt(from statement: OpaquePointer?, at index: Int32) -> Int? {
        return Int(sqlite3_column_int(statement, index))
    }
    
    //
    // function to decode OTO conversation from a json string
    func decodeOtoConversation(from jsonString: String) -> ConversationModel? {
        guard let data = jsonString.data(using: .utf8) else {
            return nil
        }
        
        do {
            let conversation = try JSONDecoder().decode(ConversationModel.self, from: data)
            return conversation
        } catch {
            print("Error decoding conversation JSON: \(error)")
            return nil
        }
    }
    
    //
    // function to decode Group conversation from a json string
    func decodeGroupConversations(from jsonString: String) -> GroupConversationModel? {
        guard let data = jsonString.data(using: .utf8) else {
            return nil
        }
        
        do {
            let conversation = try JSONDecoder().decode(GroupConversationModel.self, from: data)
            return conversation
        } catch {
            print("Error decoding group conversation JSON: \(error)")
            return nil
        }
    }
    
    
    
    //
    // function to decode Group settings from a json string
    func decodeGroupSettings(from jsonString: String) -> GroupSettingsModel? {
        guard let data = jsonString.data(using: .utf8) else {
            return nil
        }
        
        do {
            let groupSettings = try JSONDecoder().decode(GroupSettingsModel.self, from: data)
            return groupSettings
        } catch {
            print("Error decoding group settings JSON: \(error)")
            return nil
        }
    }
    
}

File: android/app/src/main/kotlin/com/transport_system/ts_driver/data_providers/database/extensions/DatabaseDecoderExtension.kt
package com.transport_system.ts_driver.data_providers.database.extensions

import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import com.transport_system.ts_driver.data_providers.models.group.GroupConversationModel
import com.transport_system.ts_driver.data_providers.models.group.GroupSettingsModel
import com.transport_system.ts_driver.data_providers.models.oto.ConversationModel
import com.transport_system.ts_driver.helpers.AppLogger


fun decodeOtoConversation(jsonString: String): ConversationModel? {
    return try {
        val gson = Gson()
        gson.fromJson(jsonString, ConversationModel::class.java)
    } catch (e: Exception) {
        println("Error decoding conversation JSON: ${e.message}")
        null
    }
}



fun decodeGroupConversation(jsonString: String): GroupConversationModel? {
    try {
        val gson = Gson()
        return gson.fromJson(jsonString, GroupConversationModel::class.java)
    } catch (e: Exception) {
        AppLogger.log("Error decoding group conversation JSON ===> ${e.message}")
        return null
    }
}


fun decodeGroupSettings(jsonString: String): GroupSettingsModel? {
    return try {
        val gson = Gson()
        gson.fromJson(jsonString, GroupSettingsModel::class.java)
    } catch (e: Exception) {
        AppLogger.log("Error decoding group settings JSON: ${e.message}")
        null  // Return null if any exception occurs
    }
}